thavalon-server/src/game/state.rs

```diff
@@ -0,0 +1,273 @@
+#![allow(dead_code)]
+use std::collections::HashSet;
+use std::fmt;
+use std::time::Duration;
+
+use super::messages::{Action, Message};
+use super::role::Team;
+use super::{Game, MissionNumber};
+
+use self::assassination::Assassination;
+use self::on_mission::{OnMission, WaitingForAgravaine};
+use self::proposing::Proposing;
+use self::voting::Voting;
+
+/// Result of handling a player action. The [`GameStateWrapper`] is the new state of the game and the [`Effect`]
+/// [`Vec`] describes side-effects of the state transition.
+pub type ActionResult = (GameStateWrapper, Vec<Effect>);
+
+/// Wrapper over specific phases, so callers can hold a game in any phase.
```

Comment:
The state machine design is roughly based on these articles and a Reddit thread about returning side-effects from state transitions:
* https://hoverbear.org/blog/rust-state-machine-pattern
* https://deislabs.io/posts/a-fistful-of-states/